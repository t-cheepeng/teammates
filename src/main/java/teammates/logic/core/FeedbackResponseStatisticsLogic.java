package teammates.logic.core;

import java.time.Instant;
import java.util.List;

import teammates.common.datatransfer.attributes.FeedbackResponseStatisticAttributes;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.storage.api.FeedbackResponseStatisticsDb;

/**
 * Handles operations related to feedback response statistics.
 *
 * @see FeedbackResponseStatisticAttributes
 * @see FeedbackResponseStatisticsDb
 */
public final class FeedbackResponseStatisticsLogic {

    private static FeedbackResponseStatisticsLogic instance = new FeedbackResponseStatisticsLogic();
    private static final FeedbackResponseStatisticsDb frsDb = new FeedbackResponseStatisticsDb();

    private FeedbackResponseStatisticsLogic() {
        // prevent initialization
    }

    /**
     * Gets the list of feedback response statistics.
     */
    public List<FeedbackResponseStatisticAttributes> getFeedbackResponseStatistics(Instant start, Instant end) {
        return frsDb.getFeedbackResponseStatistics(start, end);
    }

    /**
     * Creates a feedback response statistic with given total count.
     * If the feedback response statistic already exists, update the count.
     *
     * @return created feedback response statistic
     * @throws InvalidParametersException   if the feedback response statistic is not valid
     * @throws EntityAlreadyExistsException if the feedback response statistic already exists
     */
    public FeedbackResponseStatisticAttributes setFeedbackResponseStatistic(Instant time, int totalCount)
            throws InvalidParametersException, EntityAlreadyExistsException {
        try {
            int previousFrsTotalCount = frsDb.getPreviousFeedbackResponseStatisticTotalCount(time);
            int newCount = totalCount - previousFrsTotalCount;
            return setFeedbackResponseStatistic(time, newCount, totalCount);
        } catch (EntityDoesNotExistException e) {
            return setFeedbackResponseStatistic(time, totalCount, totalCount);
        }
    }

    private FeedbackResponseStatisticAttributes setFeedbackResponseStatistic(Instant time, int count, int totalCount)
            throws InvalidParametersException, EntityAlreadyExistsException {
        if (frsDb.hasFeedbackResponseStatistic(time)) { // already exists, increment
            int frsTotalCount = frsDb.getFeedbackResponseStatisticTotalCount(time);
            if (frsTotalCount == totalCount) {
                return FeedbackResponseStatisticAttributes.valueOf(frsDb.getFeedbackResponseStatistic(time));
            }
            return frsDb.setFeedbackResponseStatisticCount(time, count, totalCount);
        } else { // does not exist, create
            FeedbackResponseStatisticAttributes frsa = new FeedbackResponseStatisticAttributes(time, count, totalCount);
            return frsDb.createEntity(frsa);
        }
    }

    public static FeedbackResponseStatisticsLogic inst() {
        return instance;
    }
}
