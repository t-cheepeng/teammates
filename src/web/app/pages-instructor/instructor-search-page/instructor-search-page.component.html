<tm-instructor-search-bar (searched)="search($event)" [searchParams]="searchParams"></tm-instructor-search-bar>

<tm-student-result-table *ngIf="studentTables && studentTables.length" [studentTables]="studentTables">
</tm-student-result-table>
<tm-comment-result-table *ngIf="commentTables && commentTables.length" [commentTables]="commentTables">
</tm-comment-result-table>
