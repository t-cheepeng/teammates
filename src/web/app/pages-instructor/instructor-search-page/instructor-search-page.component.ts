import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { finalize } from 'rxjs/operators';
import { LoadingBarService } from '../../../services/loading-bar.service';
import { InstructorSearchResult, SearchService, } from '../../../services/search.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { ErrorMessageOutput } from '../../error-message-output';
import { StudentListSectionData } from '../student-list/student-list-section-data';
import { CommentListSectionData } from "./comment-result-table/comment-result-table.component";

/**
 * Data object for communication with the child student result component
 */
export interface SearchStudentsTable {
  courseId: string;
  sections: StudentListSectionData[];
}

/**
 * Data object for communication with child comment result component
 */
export interface SearchCommentsTable {
  courseId: string;
  sessionName: string;
  sections: CommentListSectionData[];
}

/**
 * Parameters inputted by user to be used in search
 */
export interface SearchParams {
  searchKey: string;
  isSearchForStudents: boolean;
  isSearchForComments: boolean;
}

/**
 * Instructor search page.
 */
@Component({
  selector: 'tm-instructor-search-page',
  templateUrl: './instructor-search-page.component.html',
  styleUrls: ['./instructor-search-page.component.scss'],
})
export class InstructorSearchPageComponent implements OnInit {

  searchParams: SearchParams = {
    searchKey: '',
    isSearchForStudents: true,
    isSearchForComments: false,
  }
  studentTables: SearchStudentsTable[] = [];
  commentTables: SearchCommentsTable[] = [];

  constructor(
    private route: ActivatedRoute,
    private statusMessageService: StatusMessageService,
    private loadingBarService: LoadingBarService,
    private searchService: SearchService,
  ) {}

  ngOnInit(): void {
    this.route.queryParams.subscribe((queryParams: any) => {
      if (queryParams.studentSearchkey) {
        this.searchParams.searchKey = queryParams.studentSearchkey;
      }
      if (this.searchParams.searchKey) {
        this.search(this.searchParams);
      }
    });
  }

  /**
   * Searches for students and questions/responses/comments matching the search query.
   */
  search(searchParams: SearchParams): void {
    this.loadingBarService.showLoadingBar();
    this.searchService
      .searchInstructor(searchParams.searchKey, searchParams.isSearchForStudents, searchParams.isSearchForComments)
      .pipe(finalize(() => this.loadingBarService.hideLoadingBar()))
      .subscribe(
        (resp: InstructorSearchResult) => {
          this.commentTables = resp.searchCommentsTables;
          this.studentTables = resp.searchStudentsTables;
          const hasStudents: boolean = !!(
            this.studentTables && this.studentTables.length
          );
          const hasComments: boolean = !!(
            this.commentTables && this.commentTables.length
          );
          if (!hasStudents && !hasComments) {
            this.statusMessageService.showWarningMessage('No results found.');
          }
        },
        (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorMessage(resp.error.message);
        },
      );
  }
}
