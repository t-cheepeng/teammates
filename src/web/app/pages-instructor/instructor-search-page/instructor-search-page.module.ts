import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { StudentListModule } from '../student-list/student-list.module';
import { InstructorSearchBarComponent } from './instructor-search-bar/instructor-search-bar.component';
import { InstructorSearchPageComponent } from './instructor-search-page.component';
import { StudentResultTableComponent } from './student-result-table/student-result-table.component';
import { CommentResultTableComponent } from "./comment-result-table/comment-result-table.component";
import { CommentBoxModule } from "../../components/comment-box/comment-box.module";

/**
 * Module for instructor search page.
 */
@NgModule({
  declarations: [
    InstructorSearchPageComponent,
    InstructorSearchBarComponent,
    StudentResultTableComponent,
    CommentResultTableComponent,
  ],
  exports: [
    InstructorSearchPageComponent,
    InstructorSearchBarComponent,
    StudentResultTableComponent,
    CommentResultTableComponent,
  ],
  imports: [
    CommonModule,
    StudentListModule,
    FormsModule,
    RouterModule,
    NgbModule,
    CommentBoxModule,
  ],
})
export class InstructorSearchPageModule { }
