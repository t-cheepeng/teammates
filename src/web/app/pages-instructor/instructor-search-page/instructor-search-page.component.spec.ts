import { HttpClientTestingModule } from '@angular/common/http/testing';

import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBarModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { JoinState } from '../../../types/api-output';
import { InstructorSearchPageComponent } from './instructor-search-page.component';
import { InstructorSearchPageModule } from './instructor-search-page.module';

describe('InstructorSearchPageComponent', () => {
  let component: InstructorSearchPageComponent;
  let fixture: ComponentFixture<InstructorSearchPageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule,
        RouterTestingModule,
        MatSnackBarModule,
        InstructorSearchPageModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorSearchPageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a search key', () => {
    component.searchParams.searchKey = 'TEST';
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with a student table', () => {
    component.studentTables = [
      {
        courseId: 'test.tes-demo',
        sections: [
          {
            sectionName: 'Tutorial Group 1',
            isAllowedToViewStudentInSection: true,
            isAllowedToModifyStudent: true,
            students: [
              {
                name: 'Alice Betsy',
                email: 'alice.b.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
              {
                name: 'Benny Charles',
                email: 'benny.c.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
              {
                name: 'Danny Engrid',
                email: 'danny.e.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
              {
                name: 'Emma Farrell',
                email: 'emma.f.tmms@gmail.tmt',
                status: JoinState.JOINED,
                team: 'Team 1',
              },
            ],
          },
        ],
      },
    ];

    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
