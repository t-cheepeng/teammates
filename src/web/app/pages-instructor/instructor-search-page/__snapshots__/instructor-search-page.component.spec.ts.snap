// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorSearchPageComponent should snap with a search key 1`] = `
<tm-instructor-search-page
  commentTables={[Function Array]}
  loadingBarService={[Function LoadingBarService]}
  route={[Function ActivatedRoute]}
  searchParams={[Function Object]}
  searchService={[Function SearchService]}
  statusMessageService={[Function StatusMessageService]}
  studentTables={[Function Array]}
>
  <tm-instructor-search-bar
    ng-reflect-search-params="[object Object]"
  >
    <div
      class="card bg-light mb-4"
    >
      <div
        class="card-body"
      >
        <div
          class="text-muted"
        >
           Search Tips:
          <br />
          <ul>
            <li>
              Put more keywords to search for more precise results.
            </li>
            <li>
              Put quotation marks around words 
              <b>
                "[any word]"
              </b>
               to search for an exact phrase in an exact order.
            </li>
          </ul>
        </div>
        <div
          class="form-group"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              ng-reflect-model="TEST"
              placeholder="Your search keyword"
              type="text"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
              >
                Search
              </button>
            </div>
            <div
              class="input-group mt-md-1"
            >
              <div
                class="col-md-3"
              >
                <input
                  class="form-check-inline ng-untouched ng-pristine ng-valid"
                  id="studentsCheckbox"
                  ng-reflect-model="true"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="studentsCheckbox"
                >
                  <b>
                    Students
                  </b>
                </label>
              </div>
              <div
                class="col-md-9"
              >
                <input
                  class="form-check-inline ng-untouched ng-pristine ng-valid"
                  id="commentCheckbox"
                  ng-reflect-model="false"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="commentCheckbox"
                >
                  <b>
                    Questions, responses, comments on responses
                  </b>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-instructor-search-bar>
</tm-instructor-search-page>
`;

exports[`InstructorSearchPageComponent should snap with a student table 1`] = `
<tm-instructor-search-page
  commentTables={[Function Array]}
  loadingBarService={[Function LoadingBarService]}
  route={[Function ActivatedRoute]}
  searchParams={[Function Object]}
  searchService={[Function SearchService]}
  statusMessageService={[Function StatusMessageService]}
  studentTables={[Function Array]}
>
  <tm-instructor-search-bar
    ng-reflect-search-params="[object Object]"
  >
    <div
      class="card bg-light mb-4"
    >
      <div
        class="card-body"
      >
        <div
          class="text-muted"
        >
           Search Tips:
          <br />
          <ul>
            <li>
              Put more keywords to search for more precise results.
            </li>
            <li>
              Put quotation marks around words 
              <b>
                "[any word]"
              </b>
               to search for an exact phrase in an exact order.
            </li>
          </ul>
        </div>
        <div
          class="form-group"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              ng-reflect-model=""
              placeholder="Your search keyword"
              type="text"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
              >
                Search
              </button>
            </div>
            <div
              class="input-group mt-md-1"
            >
              <div
                class="col-md-3"
              >
                <input
                  class="form-check-inline ng-untouched ng-pristine ng-valid"
                  id="studentsCheckbox"
                  ng-reflect-model="true"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="studentsCheckbox"
                >
                  <b>
                    Students
                  </b>
                </label>
              </div>
              <div
                class="col-md-9"
              >
                <input
                  class="form-check-inline ng-untouched ng-pristine ng-valid"
                  id="commentCheckbox"
                  ng-reflect-model="false"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="commentCheckbox"
                >
                  <b>
                    Questions, responses, comments on responses
                  </b>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-instructor-search-bar><tm-student-result-table
    ng-reflect-student-tables="[object Object]"
  >
    <div
      class="card border border-primary mb-4"
    >
      <div
        class="card-header text-white bg-primary font-weight-bold"
      >
        Students
      </div>
      <div
        class="card-body p-3"
      >
        
        <div
          class="card border border-light-blue mb-3"
        >
          <div
            class="card-header text-dark-blue bg-light-blue font-weight-bold"
          >
            [test.tes-demo]
          </div>
          <div
            class="card-body p-0"
          >
            <tm-student-list
              ng-reflect-course-id="test.tes-demo"
              ng-reflect-enable-remind-button="false"
              ng-reflect-sections="[object Object]"
              ng-reflect-use-gray-heading="true"
            >
              <div
                class="table-responsive"
              >
                <table
                  class="table table-bordered table-striped m-0"
                >
                  <thead
                    class="thead-gray"
                    ng-reflect-ng-class="[object Object]"
                  >
                    <tr>
                      <th>
                        Photo
                      </th>
                      
                      <th
                        class="sortable-header"
                      >
                         Section 
                        
                        <span>
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                        
                        
                      </th>
                      <th
                        class="sortable-header"
                      >
                         Team 
                        
                        <span>
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                        
                        
                      </th>
                      <th
                        class="sortable-header"
                      >
                         Student Name 
                        
                        <span>
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                        
                        
                      </th>
                      <th
                        class="sortable-header"
                      >
                         Status 
                        
                        <span>
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                        
                        
                      </th>
                      <th
                        class="sortable-header"
                      >
                         Email 
                        
                        <span>
                          <i
                            class="fas fa-sort"
                          />
                        </span>
                        
                        
                      </th>
                      <th>
                        Action(s)
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    
                    <tr
                      s=""
                    >
                      <td
                        class="align-center"
                      >
                        <tm-view-photo-popover
                          ng-reflect-is-view-photo-link-in-popover="false"
                          ng-reflect-photo-url="http://localhost:8080/webapi/s"
                        >
                          
                          
                          
                          <a
                            href="/"
                            ng-reflect-query-params-handling="preserve"
                            queryparamshandling="preserve"
                            role="button"
                          >
                             View Photo 
                          </a>
                          
                          
                          
                        </tm-view-photo-popover>
                      </td>
                      
                      <td>
                        Tutorial Group 1
                      </td>
                      <td>
                        Team 1
                      </td>
                      <td>
                        Alice Betsy
                      </td>
                      <td
                        class="align-center"
                      >
                        Joined
                      </td>
                      <td>
                        alice.b.tmms@gmail.tmt
                      </td>
                      <td
                        class="no-print align-center"
                      >
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/details?courseid=test.tes-demo&studentemail=alice.b.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View the details of the studen"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           View
                        </a>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/edit?courseid=test.tes-demo&studentemail=alice.b.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Use this to edit the details o"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           Edit
                        </a>
                        
                        
                        
                        <button
                          class="btn btn-light btn-sm mr-2"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Delete the student and the cor"
                        >
                          Delete
                        </button>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/students/records?courseid=test.tes-demo&studentemail=alice.b.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View all data about this stude"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/students/recor"
                        >
                           All Records
                        </a>
                        
                      </td>
                    </tr>
                    <tr
                      s=""
                    >
                      <td
                        class="align-center"
                      >
                        <tm-view-photo-popover
                          ng-reflect-is-view-photo-link-in-popover="false"
                          ng-reflect-photo-url="http://localhost:8080/webapi/s"
                        >
                          
                          
                          
                          <a
                            href="/"
                            ng-reflect-query-params-handling="preserve"
                            queryparamshandling="preserve"
                            role="button"
                          >
                             View Photo 
                          </a>
                          
                          
                          
                        </tm-view-photo-popover>
                      </td>
                      
                      <td>
                        Tutorial Group 1
                      </td>
                      <td>
                        Team 1
                      </td>
                      <td>
                        Benny Charles
                      </td>
                      <td
                        class="align-center"
                      >
                        Joined
                      </td>
                      <td>
                        benny.c.tmms@gmail.tmt
                      </td>
                      <td
                        class="no-print align-center"
                      >
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/details?courseid=test.tes-demo&studentemail=benny.c.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View the details of the studen"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           View
                        </a>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/edit?courseid=test.tes-demo&studentemail=benny.c.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Use this to edit the details o"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           Edit
                        </a>
                        
                        
                        
                        <button
                          class="btn btn-light btn-sm mr-2"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Delete the student and the cor"
                        >
                          Delete
                        </button>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/students/records?courseid=test.tes-demo&studentemail=benny.c.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View all data about this stude"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/students/recor"
                        >
                           All Records
                        </a>
                        
                      </td>
                    </tr>
                    <tr
                      s=""
                    >
                      <td
                        class="align-center"
                      >
                        <tm-view-photo-popover
                          ng-reflect-is-view-photo-link-in-popover="false"
                          ng-reflect-photo-url="http://localhost:8080/webapi/s"
                        >
                          
                          
                          
                          <a
                            href="/"
                            ng-reflect-query-params-handling="preserve"
                            queryparamshandling="preserve"
                            role="button"
                          >
                             View Photo 
                          </a>
                          
                          
                          
                        </tm-view-photo-popover>
                      </td>
                      
                      <td>
                        Tutorial Group 1
                      </td>
                      <td>
                        Team 1
                      </td>
                      <td>
                        Danny Engrid
                      </td>
                      <td
                        class="align-center"
                      >
                        Joined
                      </td>
                      <td>
                        danny.e.tmms@gmail.tmt
                      </td>
                      <td
                        class="no-print align-center"
                      >
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/details?courseid=test.tes-demo&studentemail=danny.e.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View the details of the studen"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           View
                        </a>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/edit?courseid=test.tes-demo&studentemail=danny.e.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Use this to edit the details o"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           Edit
                        </a>
                        
                        
                        
                        <button
                          class="btn btn-light btn-sm mr-2"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Delete the student and the cor"
                        >
                          Delete
                        </button>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/students/records?courseid=test.tes-demo&studentemail=danny.e.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View all data about this stude"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/students/recor"
                        >
                           All Records
                        </a>
                        
                      </td>
                    </tr>
                    <tr
                      s=""
                    >
                      <td
                        class="align-center"
                      >
                        <tm-view-photo-popover
                          ng-reflect-is-view-photo-link-in-popover="false"
                          ng-reflect-photo-url="http://localhost:8080/webapi/s"
                        >
                          
                          
                          
                          <a
                            href="/"
                            ng-reflect-query-params-handling="preserve"
                            queryparamshandling="preserve"
                            role="button"
                          >
                             View Photo 
                          </a>
                          
                          
                          
                        </tm-view-photo-popover>
                      </td>
                      
                      <td>
                        Tutorial Group 1
                      </td>
                      <td>
                        Team 1
                      </td>
                      <td>
                        Emma Farrell
                      </td>
                      <td
                        class="align-center"
                      >
                        Joined
                      </td>
                      <td>
                        emma.f.tmms@gmail.tmt
                      </td>
                      <td
                        class="no-print align-center"
                      >
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/details?courseid=test.tes-demo&studentemail=emma.f.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View the details of the studen"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           View
                        </a>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/courses/student/edit?courseid=test.tes-demo&studentemail=emma.f.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Use this to edit the details o"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/courses/studen"
                        >
                           Edit
                        </a>
                        
                        
                        
                        <button
                          class="btn btn-light btn-sm mr-2"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="Delete the student and the cor"
                        >
                          Delete
                        </button>
                        
                        
                        <a
                          class="btn btn-light btn-sm mr-2"
                          href="/web/instructor/students/records?courseid=test.tes-demo&studentemail=emma.f.tmms@gmail.tmt"
                          ng-reflect-klass="btn btn-light btn-sm mr-2"
                          ng-reflect-ng-class="[object Object]"
                          ng-reflect-ngb-tooltip="View all data about this stude"
                          ng-reflect-query-params="[object Object]"
                          ng-reflect-router-link="/web/instructor/students/recor"
                        >
                           All Records
                        </a>
                        
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </tm-student-list>
          </div>
        </div>
      </div>
    </div>
  </tm-student-result-table>
</tm-instructor-search-page>
`;

exports[`InstructorSearchPageComponent should snap with default fields 1`] = `
<tm-instructor-search-page
  commentTables={[Function Array]}
  loadingBarService={[Function LoadingBarService]}
  route={[Function ActivatedRoute]}
  searchParams={[Function Object]}
  searchService={[Function SearchService]}
  statusMessageService={[Function StatusMessageService]}
  studentTables={[Function Array]}
>
  <tm-instructor-search-bar
    ng-reflect-search-params="[object Object]"
  >
    <div
      class="card bg-light mb-4"
    >
      <div
        class="card-body"
      >
        <div
          class="text-muted"
        >
           Search Tips:
          <br />
          <ul>
            <li>
              Put more keywords to search for more precise results.
            </li>
            <li>
              Put quotation marks around words 
              <b>
                "[any word]"
              </b>
               to search for an exact phrase in an exact order.
            </li>
          </ul>
        </div>
        <div
          class="form-group"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-valid"
              ng-reflect-model=""
              placeholder="Your search keyword"
              type="text"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
              >
                Search
              </button>
            </div>
            <div
              class="input-group mt-md-1"
            >
              <div
                class="col-md-3"
              >
                <input
                  class="form-check-inline ng-untouched ng-pristine ng-valid"
                  id="studentsCheckbox"
                  ng-reflect-model="true"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="studentsCheckbox"
                >
                  <b>
                    Students
                  </b>
                </label>
              </div>
              <div
                class="col-md-9"
              >
                <input
                  class="form-check-inline ng-untouched ng-pristine ng-valid"
                  id="commentCheckbox"
                  ng-reflect-model="false"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="commentCheckbox"
                >
                  <b>
                    Questions, responses, comments on responses
                  </b>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </tm-instructor-search-bar>
</tm-instructor-search-page>
`;
